lib/r10k/source/exec.rb

```diff
@@ -0,0 +1,53 @@
+require 'r10k/util/subprocess'
+require 'json'
+require 'yaml'
+
+class R10K::Source::Exec < R10K::Source::Hash
+  R10K::Source.register(:exec, self)
+
+  def initialize(name, basedir, options = {})
+    unless @command = options[:command]
+      raise ConfigError, _('Environment source %{name} missing required parameter: command') % {name: name}
+    end
+
+    # We haven't set the environments option yet. We will do that by
+    # overloading the #environments method.
+    super(name, basedir, options)
+  end
+
+  def environments
+    if @environments.nil?
```

Comment:
Non-blocker: `R10K::Source::Hash` has `#environments` which lazily sets `@environments` from `#environments_hash`. Instead of overriding `#environments` we can override `#environments_hash`, which drops the need for the conditional call to `super`.

This is a good candidate for a refactor in a second PR, so we can skip this work in this PR.

Reply:
Ended up implementing this suggestion along with the hard blocker fixes